### PR TITLE
Add E2E test for Premium themes in with-theme onboarding flow

### PR DESCRIPTION
**Pull request by BogdanUngureanu**
This adds a new e2e test that covers user creation, onboarding, theme installation and account deletion for the with-theme flow with a premium theme.

Currently blocked by the changes introduced in https://github.com/Automattic/wp-calypso/pull/83863. 

Related to https://github.com/Automattic/dotcom-forge/issues/4148

## Proposed Changes

* Adds a new e2e test for a premium theme using the `with-theme` flow

## Testing Instructions

* N/A - Check the PR's e2e tests or run the test locally following the instructions in the [docs](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/README.md#quick-start).
* To execute the test for a given URL, execute it with the `CALYPSO_BASE_URL=https://calypso live link from the PR`: e.g. 
``CALYPSO_BASE_URL=https://container-tender-burnell.calypso.live/ yarn workspace wp-e2e-tests test -- test/e2e/specs/onboarding/signup__with-theme-premium.ts
``

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts]: **question:blocking** - I have a question here, and a suggestion conditional on the response to the question:

1. is it intention to remove the wait for the `waitForURL`, which adds a certain level of delay before this method hands off control.
2. if the above is intentional, a suggestion would be to remove the `Promise.all` wrapper because the only wait is for `waitForLoadState`.

File: packages/calypso-e2e/src/lib/pages/themes-detail-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/themes-detail-page.ts]: **suggestion:blocking** - FYI, we no longer suggest adding new CSS selectors unless absolutely necessary, keeping with [upstream recommendation](https://playwright.dev/docs/locators#locate-by-css-or-xpath).

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (worldomonation) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]:
```suggestion
describe( 'Lifecyle: Premium theme signup, onboard, launch and cancel subscription', function () {
```

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (worldomonation) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: **nit:non-blocking** - should be changed to something different, because this prefix isn't accurate.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (worldomonation) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: **suggestion:non-blocking** - I'm thinking we can try another currency. GBP has a lot of coverage, and while different currency shouldn't make a difference in theory, it might be helpful. Something like `DKK` or `CAD`.

If you don't want to make the change here, that's fine too - I can do that in a follow-up.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (worldomonation) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts:91]: **suggestion:non-blocking**- I'm going to actually suggest that, to make this spec more useful, we can cover the scenario of explicitly selecting a Free WordPress.com domain. What do you think?

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (worldomonation) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: **suggestion:non-blocking** - conversely, we already test this in the other flow that you copied this from, so I don't think this is really necessary. Up to you if you'd like to remove or keep it, doesn't hurt.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (worldomonation) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts:142]: **question:kinda-blocking** - we aren't checking on the UI whether the theme that was selected is actually applied. Is that "enough" testing?

File: packages/calypso-e2e/src/lib/pages/themes-detail-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/themes-detail-page.ts:84]: **question** - this is happening when the theme detail page is viewed while logged out, correct?

**Review by worldomonation — CHANGES_REQUESTED** [review]
Overall, pretty well written! I left a bunch of questions and suggestions, but half of them are non-blocking (meaning I won't block a merge on that even if you don't address it). 

The blocking ones though, I did leave suggestions for addressing them.

File: packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts
Comment (BogdanUngureanu) [reply packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts]: Yes, it's intentional because the plans page for the `with-theme` is different (`/start/with-theme/plans-theme-preselected?theme=theme-name`).

Good point. There's no reason to keep it in the Promise.all.

Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts]: > Yes, it's intentional 

Gotcha. In that case, you can either use a regex (eg. `(start\/plans|start\/with-theme)` to add some verification to the method, or just remove it entirely. 

I'm usually in favor of making sure the right URLs are being loaded, but I'll leave it up to you.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (BogdanUngureanu) [reply test/e2e/specs/onboarding/signup__with-theme-premium.ts:91]: I avoided selecting a domain name because I didn't want to risk actually registering an actual domain name.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (BogdanUngureanu) [reply test/e2e/specs/onboarding/signup__with-theme-premium.ts:142]: I think so, we are interested in testing only the onboarding flow. We want to see that the API returns the correct theme for the site.

The UI for the active theme will change anyway soon.

File: packages/calypso-e2e/src/lib/pages/themes-detail-page.ts
Comment (BogdanUngureanu) [reply packages/calypso-e2e/src/lib/pages/themes-detail-page.ts:84]: Yes :)

**Review by worldomonation — APPROVED** [review]

**Comment by worldomonation** [comment]
You seem to have permanently failing specs in the Calypso E2E, which should be dealt with before merging.

They don't seem related to your changes though, so my first suggestion would be to rebase your branch. If you are unsure how to do so, I wrote a guide on not review-bombing: p7H4VZ-3Oh-p2

**Review by dsas — APPROVED** [review]
This all makes sense to me...not that I've really looked too hard at e2e tests before

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: Can we _know_ whether the upsell will be present or not?

In general, I'd encourage to avoid this kind of validation because:
1. It will likely still throw the timeout error in the trace file, which can be misleading when inspecting.
2. It can cause the test to be flaky if the timeout isn't sufficient for some reason, and timings in E2E tests can vary depending on the current CI mood. 🌦️

Comment (BogdanUngureanu) [reply test/e2e/specs/onboarding/signup__with-theme-premium.ts]: Hi @WunderBart, the upsell always appears for Premium plans, which what we always select for Premium themes in this test.

Maybe I should rename the test to "Skip upsell to Business plan" to be less confusing.

File: packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts]: No need to wait for the load state explicitly since [it's internally done](https://playwright.dev/docs/api/class-frame#frame-wait-for-url-option-wait-until) by the `waitForURL` method.

File: packages/calypso-e2e/src/lib/pages/themes-page.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/lib/pages/themes-page.ts]: Same as above, `goto` waits for the load state as well, so this line can be removed. Also, waiting until `networkidle` is discouraged as we should rely on web assertions instead. We're probably fine with simply removing it, though :)

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: I'm not really sure what is happening here. Why are we trying to click a button that we've already waited for? Is this button possibly disabled or not rendered at all?

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: As above, this line is not necessary.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: `waitForSelector` is deprecated - let's use `await locator(selector).waitFor()` instead.

**Review by WunderBart — CHANGES_REQUESTED** [review]

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (BogdanUngureanu) [reply test/e2e/specs/onboarding/signup__with-theme-premium.ts]: Good point. This test was built in top of another test. [There](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/onboarding/lifecycle__signup-onboarding-launch-cancel.ts#L128), we have this condition because we display the upsell for Business on Premium purchases. We probably catch it because we might cover non-premium purchases where the upsell is not present.

However, in our case, we always test a Premium subscription, so I guess we can remove the try-catch.

Comment (WunderBart) [reply test/e2e/specs/onboarding/signup__with-theme-premium.ts]: Nice! Let's also remove the `waitForSelector` line above - the `click()` waits until the element is clickable.

File: packages/calypso-e2e/src/lib/components/domain-search-component.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/lib/components/domain-search-component.ts]: Can we use anything on the page (or the current URL) to determine the current case? I.e. whether it's the `with-theme` flow or something else? Wrapping the locator with try/catch is generally bad practice as we rely solely on a timeout value which can produce flaky results.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts:69]: Just to make sure, this selects the first premium theme available, correct? I assume we don't want to select any specific theme here because they often change?

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: nit:

```suggestion
			themeSlug = pageMatch?.[ 1 ] || null;
```

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: This should not be necessary, as we await the "Customize this design" locator below.

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (WunderBart) [inline test/e2e/specs/onboarding/signup__with-theme-premium.ts]: Let's use the recommended locator for text instead:

```suggestion
			await page.getByText( 'Customize this design' ).waitFor();
```

**Review by WunderBart — CHANGES_REQUESTED** [review]
Thanks for addressing the feedback! Just a few more comments 🙏

File: packages/calypso-e2e/src/lib/components/domain-search-component.ts
Comment (BogdanUngureanu) [reply packages/calypso-e2e/src/lib/components/domain-search-component.ts]: The domain-search-component is a generic class that's used in other flows, not just in the with-theme flow. One other option would be to add some kind of `skipContinueButton` argument in the method. I'm not sure if it makes sense though (that was my initial approach).

File: test/e2e/specs/onboarding/signup__with-theme-premium.ts
Comment (BogdanUngureanu) [reply test/e2e/specs/onboarding/signup__with-theme-premium.ts:69]: Yup, that's right. The themes often change in the recommended section.

File: packages/calypso-e2e/src/lib/components/domain-search-component.ts
Comment (WunderBart) [reply packages/calypso-e2e/src/lib/components/domain-search-component.ts]: Ah, I can see that below we use Promise.race for the `Continue` button click vs the target URL. In that case, this entire try/catch block can be removed! 😄 
```suggestion
```

Comment (BogdanUngureanu) [reply packages/calypso-e2e/src/lib/components/domain-search-component.ts]: That check was added recently here. https://github.com/Automattic/wp-calypso/pull/83886/files#r1388907588

Apparently, it was added because there was a delay in the UI.

cc @jdc91

Comment (WunderBart) [reply packages/calypso-e2e/src/lib/components/domain-search-component.ts]: Thanks for the context. From the stability perspective, this part doesn't really make sense since it doesn't address any delays in the UI - it's already handled by the button `click` below. If anything, we might need to extend the timeout on that click, but nothing else should be necessary here.

**Review by WunderBart — APPROVED** [review]
Looks good, thank you! 🚀

**Pull request merged**
